### PR TITLE
[docker] feat: Add GB200 (aarch64/Blackwell) Docker image and training example

### PR DESCRIPTION
**Pull request by kaixih**
# PR: Add ARM64 (GB200) Docker Image Support with SGLang

> **Note:** This is initial experimental support for GB200 — more models and configurations will follow in subsequent PRs.

## What changed

**`docker/Dockerfile.stable.sglang`** — 4 fixes to make the image work on GB200 (aarch64):

1. Pin `transformers==4.57.1` — SGLang requires <5.0; latest transformers 5.x breaks its `AutoImageProcessor.register()` API
2. Add `cachetools` — missing dependency causing `ModuleNotFoundError` at startup
3. Pin `trl<0.15` — trl 0.15+ removed `AutoModelForCausalLMWithValueHead` which verl's FSDP worker imports
4. Pin `nvidia-cudnn-cu12==9.16.0.29` — fixes a compatibility bug between cuDNN 9.10 and torch 2.9.1+cu129

**`examples/grpo_trainer/run_qwen3-8b_gb200.sh`** — new training script adapted for GB200.

## Why

The base image `lmsysorg/sglang:v0.5.6.post2` already has an ARM64 manifest, so no cross-compilation is needed. The 4 Dockerfile fixes are the only changes required to make it run on GB200.

## GB200-specific training config (Qwen3-8B + FSDP + SGLang on 4x B200)

Three things must be set differently from the standard x86 config:

```bash
# 1. Blackwell (SM=100) does not support FlashAttention v3 — use flashinfer
actor_rollout_ref.rollout.name=sglang
actor_rollout_ref.rollout.enforce_eager=True
+actor_rollout_ref.rollout.engine_kwargs.sglang.attention_backend=flashinfer

# 2. Docker --privileged bypasses NVIDIA runtime GPU injection — tell Ray explicitly
+ray_kwargs.ray_init.num_gpus=4

# 3. FSDP actor defaults to float32 — must set bfloat16 for Flash Attention
actor_rollout_ref.actor.fsdp_config.model_dtype=bfloat16
```

## Validation

Ran Qwen3-8B GRPO on GSM8K (4x B200, single node):

**Accuracy** (validated every 5 steps):

| Step | GSM8K Accuracy |
|------|---------------|
| 0 (baseline) | 25.3% |
| 5 | 32.4% |
| 10 | 39.5% |
| 15 | 44.4% |
| 20 | 48.8% |
| 25 | 57.6% |
| 30 | 59.6% |

**Performance** (steady state):
- Step time: ~400s (~6.7 min)
  - Rollout: ~265s (66%)
  - Actor update: ~84s (21%)
- GPU memory: **87.7GB / 192GB** (46% of 192GB)
- MFU: **25.4%**

cc @ISEEKYAN

### CODE REVIEW

File: docker/Dockerfile.sglang.gb200 [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+#
+# GB200-specific changes vs Dockerfile.stable.sglang:
+#   1. transformers==4.57.1  — pin instead of --upgrade; SGLang requires <5.0
+#   2. cachetools             — missing dependency in base sglang image
+#   3. trl<0.15               — trl 0.15+ removed AutoModelForCausalLMWithValueHead
+#
+# Build:
+#   docker build -f docker/Dockerfile.sglang.gb200 -t verl:sglang-gb200 .
+
+FROM lmsysorg/sglang:v0.5.6.post2
```
Comment (wuxibin89) [inline docker/Dockerfile.sglang.gb200]: Please upgrade to sglang==0.5.9

Comment (kaixih) [reply docker/Dockerfile.sglang.gb200]: Done. PTAL.

**Comment by kaixih** [comment]
### What was fixed

Several existing tests hardcoded 8-GPU assumptions or contained logic that deadlocked on a 4-GPU node. Tests run inside the GB200 container with backend-specific filters (NPU/CPU/TRT-LLM/reward model tests excluded).

**SGLang container** (`verl:sglang-arm64`):
```
215 passed, 7 skipped, 24 warnings in 471.92s (0:07:51)
```

**vLLM container** (`verl:vllm-arm64`):
```
186 passed, 10 skipped, 22 warnings in 605.02s (0:10:05)
```

### Repro

```bash
# Inside the container (SGLang)
cd /workspace/verl
pip install pytest-asyncio cupy-cuda12x mlflow-skinny -q --ignore-installed blinker

pytest -s \
  --ignore-glob="*on_npu.py" \
  --ignore-glob="*on_cpu.py" \
  --ignore-glob="*test_special_*.py" \
  --ignore-glob="*_hf_rollout*" \
  --ignore-glob="tests/special*" \
  --ignore-glob="tests/experimental" \
  --ignore-glob="tests/workers/reward_model" \
  --ignore-glob="tests/models/" \
  --ignore-glob="tests/workers/rollout/rollout_trtllm" \
  --ignore-glob="*test_vllm*" \
  --ignore-glob="*test_shared_memory*" \
  --ignore-glob="*test_bucketed_weight_transfer*" \
  tests/

# Inside the container (vLLM) — swap last three ignores:
#  remove: *test_vllm*, *test_shared_memory*, *test_bucketed_weight_transfer*
#  add:    *_sglang*
```

File: docker/Dockerfile.vllm.gb200 [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+# GB200-specific changes vs Dockerfile.stable.vllm:
+#   1. vllm==0.17.0              — install via pip instead of building from source
+#   2. transformers==4.57.1      — pin; avoid >=5.0 incompatibility
+#   3. trl<0.15                  — trl 0.15+ removed AutoModelForCausalLMWithValueHead
+#   4. cachetools                — missing in some base environments
+#
+# Build:
+#   docker build -f docker/Dockerfile.vllm.gb200 -t verl:vllm-gb200 .
+
+FROM nvidia/cuda:12.9.1-devel-ubuntu24.04
```
Comment (wuxibin89) [inline docker/Dockerfile.vllm.gb200]: Can we merge `docker/Dockerfile.vllm.gb200` into `docker/Dockerfile.vllm` and maintain only one Dockerfile for different GPU arch?

Comment (kaixih) [reply docker/Dockerfile.vllm.gb200]: I think the merge would introduce many conditional branches that make the Dockerfile harder to read and maintain, especially risky for this experimental container. How about incrementally aligning Dockerfile.stable.vllm toward this GB200 file? e.g. start from

1. upgrading the base image from ubuntu22.04 to ubuntu24.04 (removing the deadsnakes PPA, which is the biggest structural difference)
2. make a few small alignments (adding wheel to the pybind11 install, --no-cache-dir to Megatron).
3. Once the two files are structurally similar and the GB200 container is more validated, the merge would only need conditionals for the real semantic differences (vLLM version, pinned transformers, trl<0.15, NVTE cudnn path, DeepEP vs cachetools) rather than also fighting layout noise.

How do you think?

Comment (wuxibin89) [reply docker/Dockerfile.vllm.gb200]: Great!
- For the vllm, we can starting from upgrading ubuntu22.04 to ubuntu24.04
- For the sglang, is it possible to merge since they use same base image?

Comment (wuxibin89) [reply docker/Dockerfile.vllm.gb200]: @kaixih vllm base image has upgraged to ubuntu24.04 in https://github.com/verl-project/verl/pull/5724, can we unify vllm x86_64/aarch64 Dockerfile now?

**Comment by kaixih** [comment]
Tested `deepseek-llm-7b-chat` + GRPO on GSM8K after rebasing to latest main. GB200 (Blackwell) results are compared against the existing H100 (Hopper) baseline.

## Settings

| Parameter | H100×8 (Hopper) | GB200×4 (Blackwell) |
|---|---|---|
| rollout backend | sglang | vllm |
| `train_batch_size` | 1024 | 2048 |
| `ppo_mini_batch_size` | 256 | 512 |
| `ppo_micro_batch_size_per_gpu` | 80 | 128 |
| `rollout.log_prob_micro_batch_size_per_gpu` | 160 | 80 |
| `ref.log_prob_micro_batch_size_per_gpu` | 160 | 160 |
| `tensor_model_parallel_size` | 2 | 1 |
| `gpu_memory_utilization` | 0.6 | 0.5 |
| `free_cache_engine` | N/A | disabled |

The larger batch size on GB200 is enabled by the larger per-GPU VRAM (184GB vs 80GB) and `enable_chunked_prefill=True`, which processes rollout sequences in chunks so KV cache usage doesn't scale with `train_batch_size`.

## Performance

| | H100×8 | GB200×4 |
|---|---|---|
| throughput | ~3000 tok/s | ~5050-5300 tok/s |
| per-GPU throughput | ~375 tok/s | ~1300 tok/s |

GB200×4 achieves **~3.5× better per-GPU efficiency** compared to H100×8.

Note: The GB200 container is ready for use. Further performance optimizations (e.g., larger batch sizes, tuning memory utilization, exploring SGLang on Blackwell, multi nodes) are planned as follow-up work.

cc. @jqwang2373 @borisfom

File: docker/Dockerfile.vllm.gb200 (hunk 2 above)
Comment (kaixih) [reply docker/Dockerfile.vllm.gb200]: Yes, I will go ahead and work on unifying the vLLM x86_64 and aarch64 Dockerfiles to integrate the GB200 support into the main one. 

Also, as a quick follow-up regarding the performance (which I just detailed in my comment below): we've looked into the numbers and they make sense. The container is now validated and ready for use!

Comment (kaixih) [reply docker/Dockerfile.vllm.gb200]: I merged the change to the vllm stable dockerfile. And also try to make this PR focus on the vllm. PTAL @wuxibin89

File: docker/Dockerfile.stable.vllm [hunk 3]
```diff
@@ -4,10 +4,13 @@ FROM nvidia/cuda:12.9.1-devel-ubuntu24.04
...
 ARG DEBIAN_FRONTEND=noninteractive
 ARG PIP_NO_CACHE_DIR=1
+ARG APT_MIRROR=""
 # PEP 668: Ubuntu 24.04 blocks system-wide pip installs; override for Docker
 ENV PIP_BREAK_SYSTEM_PACKAGES=1
 
-RUN sed -i "s@http://.*archive.ubuntu.com@https://mirrors.tuna.tsinghua.edu.cn@g" /etc/apt/sources.list.d/ubuntu.sources
+RUN if [ -n "${APT_MIRROR}" ]; then \
+    sed -i "s@http://.*archive.ubuntu.com@${APT_MIRROR}@g" /etc/apt/sources.list.d/ubuntu.sources; \
+    fi
```
Comment (kaixih) [inline docker/Dockerfile.stable.vllm:13]: Hi @wuxibin89  Previously the Dockerfile hardcoded a Tsinghua apt mirror, which broke builds outside China. This PR replaces it with an optional APT_MIRROR build arg. So, the internal pipelines building inside China should pass `--build-arg APT_MIRROR=https://mirrors.tuna.tsinghua.edu.cn`. (<-- I couldn't find the building pipeline in this repo)

File: tests/single_controller/test_split_resource_pool.py [hunk 4]
```diff
@@ -49,12 +49,13 @@ def add(self, data: DataProto):
 
 def test_split_resource_pool_with_split_size():
     ray.init()
-    # assume we have 2 nodes, with 4 GPUs each
-    global_resource_pool = RayResourcePool(process_on_nodes=[4, 4])
+    ngpus = torch.cuda.device_count()
```
Comment (ETOgaosion) [inline tests/single_controller/test_split_resource_pool.py:56]: Maybe we can extract the same codes to get ngpus and half_ngpus or quater_ngpus?

Like we can provide a `get_local_gpus_num(devisions=[1,2,4])` function?

**Comment by ETOgaosion** [comment]
If there have been published images, please add them to our docs?

**Comment by kaixih** [comment]
> If there have been published images, please add them to our docs?

Thanks for the suggestion. Pre-built GB200 (aarch64) images are not yet published. Users will need to build locally for now. We've added a note and the build command to the doc for now.

File: tests/single_controller/test_split_resource_pool.py (hunk 4 above)
Comment (kaixih) [reply tests/single_controller/test_split_resource_pool.py:56]: Done. PTAL.
